src/pymmcore_widgets/mda/_core_mda.py

```diff
@@ -197,9 +222,29 @@ def _on_run_clicked(self) -> None:
         ):
             return
 
+        sequence = self.value()
+
+        # get saving info from the metadata
+        metadata = sequence.metadata.get("pymmcore_widgets", None)
+        save_dir = metadata.get("save_dir") if metadata else None
+        save_name = metadata.get("save_name") if metadata else EXP
+        extension = metadata.get("extension") if metadata else None
+
+        # create the writers path and make sure they are unique
+        writer_path = None
+        if save_dir and extension is not None:
+            path = Path(save_dir) / f"{save_name}"
+            writer_path = ensure_unique(path, extension)
```

Comment:
because `ensure_unique` is such an opinionated function (i.e. it makes specific decisions that a user may very well want to modify), we need to make it easier to customize it.  Currently the only way they can do that is to either try to monkeypatch a private function (`_util.ensure_unique`), or subclass `MDAWidget` and re-implement a complicated private method `MDAWidget._on_run_clicked`.  So, try to think about providing a nicer way for end users to customize this behavior... even if that end-user is napari-micromanager for example.  Think about exposing very small bits of public functionality that would let someone do this themselves.  Think about what the inputs to that functionality should be and what the outputs should be.

Reply:
thanks @tlambert03!

do you mean something like a defined protocol with for example the `ensure_unique` and `get_next_available_path` methods (of course names can be different)

```py
class EnsureUniquePath(Protocol):

    @abstractmethod
    def ensure_unique(self, path: Path | str, extension: str = ".tif", ndigits: int = 3) -> Path:
        ...

    @abstractmethod
    def get_next_available_path(self, path: Path | str, extension: str, ndigits: int = 3) -> Path:
        ...
```

and then in the `MDAWidget` we can have either a method or a property to set the `EnsureUniquePath`

```py
class MDAWidget(MDASequenceWidget):
    def __init__(
        self, *, parent: QWidget | None = None, mmcore: CMMCorePlus | None = None
    ) -> None:
        ...
        self._ensure_unique: EnsureUniquePath = EnsureUniquePath()
        ...

    @property
    def unique_path(self) -> EnsureUniquePath:
        return self._ensure_unique

    @unique_path.setter
    def unique_path(self, value: EnsureUniquePath) -> None:
        self._ensure_unique = value
```

Reply:
by protocol, I don't necessarily mean it _has_ to be a literally `typing.Protocol`, unless you think the complexity of the problem deserves it for example, do you think you widgets would use *both* `ensure_unique` and `get_next_available_path`?  or is it more likely that `get_next_available_path` is just an implementation detail used by `ensure_unique?`.  If so, then perhaps it's just a single callable function.  Or, it could all be put into a public method that someone could subclass and override.

I guess I'm just saying give it a shot: come up with a proposal to make this customizeable somehow, but in as minimal of a way as possible that befits the complexity of the problem

Reply:
@tlambert03 not sure if it's the best way but I created a public the method `get_next_available_paths` that returns the next _**two**_ available path so we can use the first for the `run_mda` and the second to update the `save_name` text in the `MDAWidget`.
A user can override this method and and as long as the return type is a `list` of `Path`, it will work (even if there is only one path in the list. the only thing is that the `save_name` won't be updated ).

Reply:
I was thinking that maybe it is even worth making public the `_on_run_clicked` method or `MDASequenceWidget` and `MDAWidget`...